data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3ScanPartitionCreationSupplier.java

```diff
@@ -110,13 +113,13 @@ private List<PartitionIdentifier> listFilteredS3ObjectsForBucket(final List<Stri
                                                                      final LocalDateTime startDateTime,
                                                                      final LocalDateTime endDateTime,
                                                                      final Map<String, Object> globalStateMap) {
-        Instant mostRecentLastModifiedTimestamp = globalStateMap.get(bucket) != null ? Instant.parse((String) globalStateMap.get(bucket)) : null;
+        final Instant previousScanTime = globalStateMap.get(bucket) != null ? Instant.parse((String) globalStateMap.get(bucket)) : null;
         final List<PartitionIdentifier> allPartitionIdentifiers = new ArrayList<>();
         ListObjectsV2Response listObjectsV2Response = null;
         do {
             listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request.fetchOwner(true).continuationToken(Objects.nonNull(listObjectsV2Response) ? listObjectsV2Response.nextContinuationToken() : null).build());
             allPartitionIdentifiers.addAll(listObjectsV2Response.contents().stream()
-                    .filter(s3Object -> isLastModifiedTimeAfterMostRecentScanForBucket(bucket, s3Object, globalStateMap))
+                    .filter(s3Object -> isLastModifiedTimeAfterMostRecentScanForBucket(previousScanTime, s3Object))
```

Comment:
Should we also filter out items with a `Last-Modified` which is greater than `updatedScanTime`?

As I understand, you will start this scan at time T. You may read an object from time T+1. Then, when you scan again at T+2, you are still using T to filter. Thus, you may re-read the T+1 item.

Reply:
Aren't the items with Last-modified greater than updatedScanTime ones we also want to find? 

Re-reading is not an issue, and just means there will be an extra get call to the source coordination store. Before this PR, we were basing this filter on the objects we read in the scan, but there can be a race condition where we read some objects with a timestamp that is earlier than another that is missed in the scan due to a race condition. This implementation guarantees we won't hit this, at the cost of doing some extra lookups to the coordination store.

Reply:
I see. Because we store the exact key in the coordination store, we won't re-read the actual S3 objects, right?

Reply:
Yes that's right. The main purpose for tracking the last modified here and filtering out is to just reduce lookups to the coordination store